### PR TITLE
Open new api for external system to add, check and terminate signature

### PR DESCRIPTION
**Pull request by MichaelTsengLZ**

### CODE REVIEW

**Comment by MichaelTsengLZ** [comment]
The pull request includes:
1. Add addSignature(), hasSignature() and terminate() endpoints for other micro services.
2. Find cla can add dates conditions.

File: src/server/api/cla.js [hunk 1]
```diff
@@ -418,7 +419,7 @@ module.exports = {
                     self.validatePullRequests(tmpReq, callback);
                 };
             }), function (err) {
-                done(error);
+                done(err);
```
Comment (jeffwilcox) [inline src/server/api/cla.js]: Can't this just get replaced...

`function (err) { done(err); }`

with

`, done)` ?

File: src/server/api/cla.js [hunk 2]
```diff
@@ -438,32 +439,14 @@ module.exports = {
...
-                        if (error) {
-                            log.error(error);
-                        }
-                    });
-                } else if (item.org) {
-                    req.args.org = item.org;
-                    self.validateOrgPullRequests(req);
-                } else {
-                    self.validatePullRequests(req);
+            self.validate(req, function (er) {
```
Comment (jeffwilcox) [inline src/server/api/cla.js]: If it was me, I'd prefer a name like `validateError` over `er`...

File: src/server/api/cla.js [hunk 3]
```diff
@@ -501,6 +484,112 @@ module.exports = {
...
+            user: Joi.string().required(),
+            userId: Joi.number().required(),
+            org: Joi.string(),
+            owner: Joi.string(),
+            repo: Joi.string(),
+            custom_fields: Joi.string()
+        }).and('repo', 'owner').xor('repo', 'org');
+        Joi.validate(req.args, schema, { abortEarly: false, convert: false }, function (error) {
+            if (error) {
+                return done(error.message);
```
Comment (jeffwilcox) [inline src/server/api/cla.js]: In the code above you call `return done(err)` essentially; here you are `return done(err.message)` - is the different intentional? In one case you pass back a string as the error, in the other, you pass the error object instance.

File: src/server/api/cla.js [hunk 4]
```diff
@@ -501,6 +484,112 @@ module.exports = {
...
+        var schema = Joi.object().keys({
+            user: Joi.string().required(),
+            userId: Joi.number().required(),
+            org: Joi.string(),
+            owner: Joi.string(),
+            repo: Joi.string()
+        }).and('repo', 'owner').xor('repo', 'org');
+        Joi.validate(req.args, schema, { abortEarly: false, convert: false }, function (error) {
+            if (error) {
+                return done(error.message);
```
Comment (jeffwilcox) [inline src/server/api/cla.js]: Same - is this sending back the message, or the error?

File: src/server/api/cla.js [hunk 5]
```diff
@@ -501,6 +484,112 @@ module.exports = {
...
+            }
+            req.args.owner = req.args.owner || req.args.org;
+            delete req.args.org;
+            cla.terminate(req.args, function (err, dbCla) {
+                if (err) {
+                    log.error(err);
+                    return done(err);
+                }
+                self.validate(req, function (er) {
```
Comment (jeffwilcox) [inline src/server/api/cla.js]: This is async... shouldn't this be something like

```
return self.validate(req, function (er) {
  if (er) {
     // ..
  }
  done(null, dbCla);
}
```

File: src/server/api/cla.js [hunk 6]
```diff
@@ -501,6 +484,112 @@ module.exports = {
...
+        var self = this;
+        self.getLinkedItem({
+            args: {
+                repo: req.args.repo,
+                owner: req.args.owner
+            }
+        }, function (error, item) {
+            req.args.token = item.token;
+            req.args.gist = item.gist;
+            if (item.sharedGist) {
```
Comment (jeffwilcox) [inline src/server/api/cla.js]: Similar notes... are these async functions or not? Worried there may be timing issues.

Comment (MichaelTsengLZ) [inline src/server/api/cla.js]: All of the functions validateSharedGistItems(), validateOrgPullRequests(), validatePullRequests() are async but takes very long time to finish. So we call done before they are finished and only log errors if exists.

File: src/server/services/cla.js [hunk 7]
```diff
@@ -252,6 +257,42 @@ module.exports = function () {
...
+        var deferred = q.defer();
+        var query = {
+            user: user,
+            gist_url: gist_url,
+            org_cla: !!orgId
+        };
+        if (repoId) {
+            query.repoId = repoId;
+        }
+        if (orgId) {
```
Comment (MichaelTsengLZ) [inline src/server/services/cla.js]: No. The document DB treats query.repoId = undefined and query.repoId not exists differently.

File: src/server/services/cla.js [hunk 8]
```diff
@@ -105,73 +102,81 @@ module.exports = function () {
...
+    var getPR = function (owner, repo, number, token) {
+        var deferred = q.defer();
+        github.call({
+            obj: 'pullRequests',
+            fun: 'get',
+            arg: {
+                owner: owner,
+                repo: repo,
+                number: number,
+                noCache: true
```
Comment (MichaelTsengLZ) [inline src/server/services/cla.js]: Make sense. The TTL is 5 minutes currently.

File: src/server/api/cla.js [hunk 9]
```diff
@@ -501,6 +484,112 @@ module.exports = {
...
+            if (error) {
+                return done(error.message);
+            }
+            req.args.owner = req.args.owner || req.args.org;
+            delete req.args.org;
+            cla.check(req.args, done);
+        });
+    },
+
+    terminate: function (req, done) {
```
Comment (MichaelTsengLZ) [inline src/server/api/cla.js]: I will add this feature.

File: src/server/services/cla.js [hunk 10]
```diff
@@ -264,166 +305,149 @@ module.exports = function () {
...
-                });
-            };
-            if (!args.repoId && !args.orgId) {
-                getRepo(args, function (error, repo) {
-                    if (error || !repo) {
-                        deferred.reject();
-                        if (typeof done === 'function') {
-                            done(error);
-                        }
+                    args.gist_version = gist.history[0].version;
```
Comment (MichaelTsengLZ) [inline src/server/services/cla.js]: The function of getLastSignature() only checks whether a contributor signs the current version of a CLA gist.

File: src/server/middleware/authenticated.js [hunk 11]
```diff
@@ -22,29 +23,42 @@ function authenticateForExternalApi(req, res, next) {
...
         }
-        var hasPermission = data.permissions['push'];
-        cb(err, hasPermission);
-    });
+        if (!utils.couldBeAdmin(user.login) || (req.args.org && user.scope.indexOf('admin:org_hook') < 0)) {
+            return res.status(403).json({ message: noPermissionMsg });
+        }
+        var promises = [];
+        if (req.args.owner && req.args.repo) {
+            promises.push(utils.checkRepoPushPermissionByName(req.args.repo, req.args.owner, user.token));
```
Comment (MichaelTsengLZ) [inline src/server/middleware/authenticated.js]: I'm not sure whether I should put checkRepoPushPermissionByName() here or [the place where call the addSignature() ](https://github.com/MichaelTsengLZ/cla-assistant/pull/26/files#diff-b6b8a5a588eadce62ef8cc99bf267ebeR498).  Any idea?

File: src/server/middleware/authenticated.js [hunk 12]
```diff
@@ -23,34 +23,32 @@ function authenticateForExternalApi(req, res, next) {
...
-            return;
-        }
-        var hasPermission = data.permissions['push'];
-        cb(err, hasPermission);
-    });
-}
-
 function authenticateForAdminOnlyApi(req, res, next) {
     passport.authenticate('token', { session: false }, function (err, user) {
+        var noPermissionMsg = 'You have not enough rights to call this API';
```
Comment (jeffwilcox) [inline src/server/middleware/authenticated.js]: Consider making the error message a little more helpful, "Must have admin:org_hook permission scope" or similar like that - to help explain it

File: src/server/middleware/utils.js [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+        githubService.call({
+            obj: 'repos',
+            fun: 'get',
+            arg: {
+                repo: repo,
+                owner: owner
+            },
+            token: token
+        }, function (err, data) {
+            if (err || !data) {
```
Comment (jeffwilcox) [inline src/server/middleware/utils.js]: While this works fine, another approach would be:

```
if (!err && !data) {
  err = new Error('No data returned');
}
```

before the existing code, where you could change `if (err || !data)` to `if (err)`

File: src/server/middleware/utils.js [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+                return deferred.reject(err);
+            }
+            if (!data.permissions['push']) {
+                return deferred.reject('You have no admin permission for this repo');
```
Comment (jeffwilcox) [inline src/server/middleware/utils.js]: "You do not have admin permission for this repo" sounds a little more natural to me

**Pull request merged**